[release/8.0-staging] Fix few RandomAccess.Write edge case bugs

adamsitnik opened the pull request:
Backport of https://github.com/dotnet/runtime/pull/108380 and https://github.com/dotnet/runtime/pull/109340 and https://github.com/dotnet/runtime/pull/109826 to release/8.0-staging

## Customer Impact

- [x] Customer reported
- [x] Found internally

Customers using `RandomAccess.Write` overload that accepts multiple buffers have observed `IOException` being thrown in two scenarios:
- On Windows, when the file offset (stored as `Int32`) overflows and the method tries to perform a write with negative file offset. Reported by the user in https://github.com/dotnet/runtime/issues/108322.
- On Unix, when the number of buffers is larger than `IOV_MAX` (usually 1024). Reported by the user in https://github.com/dotnet/runtime/issues/108383.

When I was fixing  https://github.com/dotnet/runtime/issues/108383 I've realized that the logic for handling incomplete writes has a bug.
For incomplete multi-buffer writes, the API was not reporting any exceptions. It returns void, so it should write everything or throw an exception.

## Regression

- [ ] Yes
- [x] No

We had all 3 bugs since the API was introduced (.NET 6).

## Testing

New unit tests were added. They were failing before the fix were applied. They are passing now.

## Risk

Low, the fix is simple and covered with tests.

NicoAvanzDev committed:
RandomAccess.Write throws System.IO.IOException Fix (#108380)

adamsitnik committed:
Handle IOV_MAX limit in RandomAccess, add missing test (#109340)

* add test for Int32 overflow for WriteGather in RandomAccess

* add failing test fore more than IOV_MAX buffers

* fix both the native and managed parts

---------

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>
Co-authored-by: Stephen Toub <stoub@microsoft.com>

adamsitnik commented:
/azp run runtime-libraries-coreclr outerloop

adamsitnik committed:
fix the tests: use AssertExtensions.SequenceEqual to compare the contents of ROM

adamsitnik commented:
/azp run runtime-libraries-coreclr outerloop

adamsitnik commented:
The failures are related, I am going to send a follow up PR to main and sync this branch soon

jeffhandley commented:
@adamsitnik I've removed the https://github.com/dotnet/runtime/labels/servicing-consider since you're still investigating the CI failures. Once it's ready for tactics review, please reapply the label.

Does this also need to be backported to release/9.0-staging?

adamsitnik committed:
More WriteGather fixes (#109826)

* don't run these tests in parallel, as each test cases uses more than 4 GB ram and disk!

* fix the test: handle incomplete reads that should happen when we hit the max buffer limit

* incomplete write fix:

- pin the buffers only once
- when re-trying, do that only for the actual reminder

* Use native memory to get OOM a soon as we run out of memory (hoping to avoid the process getting killed on Linux when OOM happens)

* For macOS preadv and pwritev can fail with EINVAL when the total length of all vectors overflows a 32-bit integer.

* add an assert that is going to warn us if vector.Count is ever more than Int32.MaxValue

---------

Co-authored-by: Michał Petryka <35800402+MichalPetryka@users.noreply.github.com>
# Conflicts:
#	src/libraries/System.IO.FileSystem/tests/RandomAccess/WriteGatherAsync.cs

adamsitnik commented:
/azp run runtime-libraries-coreclr outerloop

carlossanlop commented:
@adamsitnik @jeffhandley friendly reminder that today's Code Complete day for the February 2025 Release. If you want this PR included in that release ,please make sure it gets a Tactics approval, a sign-off from an area owners, the CI failures are investigated, and merged before 4pm PT. Otherwise it will have to wait until next month.

carlossanlop commented:
Friendly reminder that code complete is on April 14th for the May Release. If you'd like to get this change included in that release, please get a Tactics approval and merge this PR before that date.

stephentoub commented:
@adamsitnik, @jeffhandley, should this be closed?

adamsitnik committed:
Merge branch 'release/8.0-staging' into backportWriteGather8

adamsitnik commented:
/azp run runtime-libraries-coreclr outerloop

tarekgh commented:
@adamsitnik 

Just reminder, if you want to get this in for next servicing, you need to get this merged by Monday Jul 14th 4pm PT

adamsitnik commented:
> @adamsitnik
> 
> Just reminder, if you want to get this in for next servicing, you need to get this merged by Monday Jul 14th 4pm PT

Does it mean I just need to merge it myself? Or get Build Analysis green?

adamsitnik committed:
Merge branch 'release/8.0-staging' into backportWriteGather8

tarekgh commented:
> Does it mean I just need to merge it myself (edit: in the past, Carlos was just merging all servicing PRs when the time was right)? Or get Build Analysis green?

Both. Everyone has permission to merge with staging branches, so it is the same as main.
